frontend/src/reducer/notificationPage.js

```diff
@@ -49,42 +61,220 @@ export const serializeNotification = notification => {
+  return authorList.every((author, index) => {
+    if (index === 0) return true
+    return author.userId === authorList[index - 1].userId
+  })
+}
+
+const hasSameWorkspace = workspaceList => {
+  if (workspaceList.some(workspace => !workspace)) return false
+  return workspaceList.every((workspace, index) => {
+    if (index === 0) return true
+    return workspace.id === workspaceList[index - 1].id
+  })
+}
+
+const hasSameContent = notificationList => {
+  if (notificationList.some(notification => !notification.content)) return false
+  return notificationList.every((notification, index) => {
+    if (index === 0) return true
+    const previousContentId = notificationList[index - 1].type.includes(CONTENT_TYPE.COMMENT)
+      ? notificationList[index - 1].content.parentId
+      : notificationList[index - 1].content.id
+    const contentId = notification.type.includes(CONTENT_TYPE.COMMENT)
+      ? notification.content.parentId
+      : notification.content.id
+    return contentId === previousContentId
+  })
+}
+
+const addNewNotificationGroup = (notification, newNotificationList, indexInNewList, numberOfNotificationsToGroup, numberOfCriteria = NUMBER_OF_CRITERIA.TWO) => {
+  if (
+    indexInNewList >= (numberOfNotificationsToGroup - 1) &&
+    !newNotificationList
+      .slice(indexInNewList - (numberOfNotificationsToGroup - 1), indexInNewList)
+      .some(notification => notification.group)
+  ) {
+    const previousNotificationList = newNotificationList
+      .slice(indexInNewList - (numberOfNotificationsToGroup - 1), indexInNewList)
+    const isGroupedByAuthor = hasSameAuthor([notification.author, ...previousNotificationList.map(notification => notification.author)])
+    const isGroupedByWorkspace = hasSameWorkspace([notification.workspace, ...previousNotificationList.map(notification => notification.workspace)])
+    const isGroupedByContent = hasSameContent([notification, ...previousNotificationList])
+
+    if (
+      ((numberOfCriteria === NUMBER_OF_CRITERIA.TWO &&
+        (isGroupedByContent ? (isGroupedByAuthor || isGroupedByWorkspace) : (isGroupedByAuthor && isGroupedByWorkspace))) ||
+        (numberOfCriteria === NUMBER_OF_CRITERIA.ONE && (isGroupedByContent || isGroupedByAuthor || isGroupedByWorkspace))
+      ) &&
+      (!previousNotificationList.some(notification => notification.type.includes(TLM_ET.MENTION)))
+    ) {
+      const authorList = uniqBy([
+        notification.author,
+        ...previousNotificationList.map(notification => notification.author)
+      ], 'userId')
+
+      for (let i = 0; i < (numberOfNotificationsToGroup - 1); i++) newNotificationList.pop()
+
+      newNotificationList.push({
+        author: authorList,
+        id: notification.id,
+        type: `${numberOfCriteria}${isGroupedByContent ? `.${GROUP_NOTIFICATION_CRITERIA.CONTENT}` : ''}${isGroupedByAuthor ? `.${GROUP_NOTIFICATION_CRITERIA.AUTHOR}` : ''}${isGroupedByWorkspace ? `.${GROUP_NOTIFICATION_CRITERIA.WORKSPACE}` : ''}`,
```

Comment:
Make make this an array or an object so you don't need to build and then parse a string to get back the result, making the resulting code both more efficient and readable.

You could check that a group is by content by checking groupedNotification.type.groupedByContent, or groupedNorification.type.groupedBy.includes('content') for instance.

groupedBy` could be outside the `type` field by the way